# RNAseq Analysis Pipeline

## Usage

Execute `main.sh` as follows:
```bash
$ ./main.sh \
    -i <path to read (directory)> -g <path to reference genome file> \
    [-o <output directory>] [-m <min read len>] [-M <max read len>] \
    [-b <bowtie arguments>] [-n] [-p]
```
This will create a `mapping_results` directory which contains two directories:
* `runs` stores all data related to each individual read file
* `results` contains data generated by scripts from the `scripts` folder

Note: make sure to delete old results before creating new ones.

## Extras

Additional useful scripts are contained in `extra`.

* `sequential_pipeline.sh`
  * map length-filtered reads against multiple genomes in succession
* `plot_sequential_data.py`
  * visualize data obtained from sequential pipeline
* `plot_expression_differences.py`
  * visualize differences in RNAseq expression levels over pairs of samples
* `utils.py`
  * Various helper methods

## Dependencies

Tools:
* cutadapt
* fastqc
  * [more info](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/)
* bowtie2
* samtools
  * [specifications](https://samtools.github.io/hts-specs/SAMv1.pdf)
* bedtools
* moreutils

Languages:
* bash
* python
  * numpy
  * pandas
  * seaborn
  * matplotlib
  * tqdm
  * biopython
  * pysam
  * joblib
